src/Waargonaut/Prettier.hs

```diff
@@ -0,0 +1,157 @@
+                                              _Wrapped)
+
+import           Natural                     (Natural, minus, successor', zero',
+                                              _Natural)
+
+import qualified Data.Text.Lazy              as LT
+import qualified Data.Text.Lazy.Builder      as TB
+
+import           Data.Bool                   (Bool, bool)
+import           Data.Foldable               (elem, length)
+import           Data.Function               (($))
+import           Data.Functor                (fmap)
+import           Data.Maybe                  (maybe)
+import           Data.Semigroup              ((<>))
+import           Data.Traversable            (traverse)
+import qualified Data.Vector                 as V
+
+import qualified Control.Lens                as L
+import qualified Control.Lens.Plated         as P
+
+import           Waargonaut.Encode           (Encoder, runEncoder)
+import           Waargonaut.Types.CommaSep   (Elems)
+import qualified Waargonaut.Types.CommaSep   as CS
+import           Waargonaut.Types.JObject    (HasJAssoc (..), JAssoc)
+import           Waargonaut.Types.Json       (AsJType (..), JType (..), Json,
+                                              jsonTraversal, waargonautBuilder)
+import           Waargonaut.Types.Whitespace (WS (..), Whitespace (..),
+                                              wsBuilder)
+
+-- | Some choices for how the Json is indented.
+data InlineOption
+  = ArrayOnly  -- ^ Only keep array elements on the same line, input line breaks between object values.
+  | ObjectOnly -- ^ Only keep object elements on the same line, input line breaks between array values.
+  | Both       -- ^ Keep both object and array elements on the same line.
+  | Neither    -- ^ Input line breaks for both array and object elements.
+  deriving (Show, Eq)
+
+-- | Newtype to indicate how many spaces we would like to use for the indentation
+--
+newtype NumSpaces = NumSpaces Natural
+  deriving (Eq, Show)
+
+-- | Newtype for how many spaces the indentation should be increased by for each level.
+--
+-- A safe assumption is for this value to be the same as the number of steps for
+-- the identation. Such that an indentation of two spaces will be increased by
+-- two for each subsequent level.
+--
+newtype IndentStep = IndentStep Natural
+  deriving (Eq, Show)
+
+-- | Encode an @a@ directly to a 'ByteString' using the provided 'Encoder', the
+-- output will have newlines and indentation added based on the 'InlineOption' and
+-- 'NumSpaces'.
+simpleEncodePretty
+  :: Applicative f
+  => InlineOption
+  -> IndentStep
+  -> NumSpaces
+  -> Encoder f a
+  -> a
+  -> f LT.Text
+simpleEncodePretty io step ind enc =
+  fmap (TB.toLazyText . waargonautBuilder wsBuilder . prettyJson io step ind)
+  . runEncoder enc
+
+objelems :: AsJType r WS a => Traversal' r (Elems WS (JAssoc WS a))
+objelems = _JObj . _1 . _Wrapped . CS._CommaSeparated . _2 . _Just
+
+-- I'm not sure this is a legal traversal
```

Comment:
Looks legit but I'm not sure. For reference the traversal laws can be found [here](https://hackage.haskell.org/package/lens-4.17/docs/Control-Lens-Type.html#t:Lens)